File: Defanging an IP Address/Defanging an IP Address.java
/*
Given a valid (IPv4) IP address, return a defanged version of that IP address.
A defanged IP address replaces every period "." with "[.]".

Example 1:
Input: address = "1.1.1.1"
Output: "1[.]1[.]1[.]1"

Example 2:
Input: address = "255.100.50.0"
Output: "255[.]100[.]50[.]0"

Constraints:
    1. The given address is a valid IPv4 address.
 */

/**
 * Approach 1: Replace Function of String
 * 为什么我要把一道这么简单的题目放上来呢。。。
 * 不过大家可以去看下 String 中的 replace 和 replaceAll 这两个方法的源码，
 * 相信会有所收获的（replace 实际上调用了 replaceAll 方法）
 *
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */
class Solution {
    public String defangIPaddr(String address) {
        // return address.replace(".", "[.]");
        return address.replaceAll("\\.", "[.]");
    }
}

/**
 * Approach 2: Using StringBuilder
 * 如果你认真去看了 replaceAll 函数的实现，那么你应该很容易明白为什么这个解法在速度上快了一倍。
 *
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */
class Solution {
    public String defangIPaddr(String address) {
        StringBuilder ans = new StringBuilder();
        for (char c : address.toCharArray()) {
            if (c != '.') {
                ans.append(c);
            } else {
                ans.append("[.]");
            }
        }
        return ans.toString();
    }
}

File: Defanging an IP Address/Defanging an IP Address.py
'''
Given a binary string S (a string consisting only of '0' and '1's) and a positive integer N, 
return true if and only if for every integer X from 1 to N, the binary representation of X is a substring of S.

Example 1:
Input: S = "0110", N = 3
Output: true

Example 2:
Input: S = "0110", N = 4
Output: false
 
Note:
    1. 1 <= S.length <= 1000
    2. 1 <= N <= 10^9
'''

# Approach:
# 解法详解参考同名 java 文件
class Solution:
    def defangIPaddr(self, address: str) -> str:
        return address.replace('.', '[.]')